### DISCUSSION TITLE
how to create a template message

### DISCUSSION THREAD

**Discussion by project-ASTA** [Secreto31126/whatsapp-api-js#332] [Q&A]
Hello! I came across your documentation on [sending template messages](https://github.com/Secreto31126/whatsapp-api-js/blob/main/EXAMPLES/template.md) using your WhatsApp API package. I'm interested in knowing how to create a template. Could you guide me on how to do this before sending messages via WhatsApp using your package? Thanks in advance for your help!

**Comment by Secreto31126** [comment]
In order to create a template for your business, go to https://business.facebook.com/wa/manage/message-templates (make sure you are using the right account and business), click on the `Create template` button at the top right, follow the instructions to create your template (remember the template name and language as you will need them to send it), fill the template with the text, headers and buttons you want in the message (you may add variables too), and then you just have to send it to revision.

It may take between 5 minutes to a few days until it's reviewed and you can finally start using it. If you are using a test account, the activation might be instant. After that, you may start sending the messages using the package's methods. Follow the documentation and examples to see different usages, specially if you use variables and buttons.

Hope this helps you get started! If you have any other question feel free to ask it.

**Reply by project-ASTA** [reply]
This is great, Thanks, I am already aware of this, but I'm wondering how do i create a template using this package, i assumed to use the _build() fn in the template component get the stringify data and post to  [Facebook Endpoint](https://graph.facebook.com/{{api-version}}/{{waba-id}}/message_templates) as said in their [documentation](https://developers.facebook.com/docs/whatsapp/business-management-api/message-templates),

this is what in my body(as a result of _build()):
`{
    "name": "new_one_new",
    "language": {
        "code": "en_US",
        "policy": "deterministic"
    },
    "components": [
        {
            "type": "header",
            "parameters": [
                {
                    "type": "text",
                    "text": "Hello"
                }
            ]
        },
        {
            "type": "body",
            "parameters": [
                {
                    "type": "text",
                    "text": "Hello"
                }
            ]
        }
    ]
}`

and this is the result i get:
`{
    "error": {
        "message": "(#100) Unexpected key \"parameters\" on param \"components[0]\".",
        "type": "OAuthException",
        "code": 100,
        "fbtrace_id": "AqnYM3Qx4EASy4cjKO9k7Tz"
    }
}`

i assumed there is bearer token error, but i copied the offcial testing_example_template from meta, and tried it again:

body used:
`{
  "name": "order_delivery_update",
  "language": "en_US",
  "category": "UTILITY",
  "components": [
    {
      "type": "HEADER",
      "format": "LOCATION"
    },
    {
      "type": "BODY",
      "text": "Good news {{1}}! Your order #{{2}} is on its way to the location above. Thank you for your order!",
      "example": {
        "body_text": [
          [
            "Mark",
            "566701"
          ]
        ]
      }
    },
    {
      "type": "FOOTER",
      "text": "To stop receiving delivery updates, tap the button below."
    },
    {
      "type":"BUTTONS",
      "buttons": [
        {
          "type": "QUICK_REPLY",
          "text": "Stop Delivery Updates"
        }
      ]
    }
  ]
}`

response:
`{
    "id": "280440051821830",
    "status": "APPROVED",
    "category": "UTILITY"
}`

which means my configuration is right but seems there is an error in the component structure, so could you guide me or is it possible to create a new template using this package??

**Reply by Secreto31126** [reply]
Oh, my bad. Unfortunately, the library does not provide a method for creating templates. The _build method is an internal function used to stringify the messages that will be sent to the user via the API.

For future reference, any prepended underscore method and readonly property in the documentation should be treated as package-private (aka don't use them unless you want to take the risk it might break with an update). It's one of the main things I never managed to get right, but unfortunately there's no easy way to implement such type in TS nor TypeDoc.

I try really hard to strike a good balance between an accessible documentation and a full technical one. I'm sorry if this presentation issue confused you.

**Reply by project-ASTA** [reply]
I understand, Template Creating is a Key part of Cloud API though, i used your package for sending text, media message all work good.
Is there a way you suggest me to create template other than manually creating them?

**Reply by Secreto31126** [reply]
Sorry for the late response. 

I do understand that template creation is a big deal in the API, however, I struggle to find scenarios where automation could add any extra value. In most cases the template should be human typed, translated and validated, so adding the logic to automate it seems irrational. 

Furthermore, I think any business management would not fit in the library's purpose, which is chatting. The package does provide some methods for media and QR management, but those are intended as complementary features to user interaction rather than account automation.

Back when I was starting the project I thought about writing an account manager assistant in bash. Unfortunately, due to time and logistics I never got around to creating it. Plus, a CLI might not be the best interface for template creation 😂.

TL;DR: I don't know why you would want to automate it. Let me know what you are planning and see if we can find a solution to it :)
